README.rst
==============
mariadb-dyncol
==============

.. image:: https://img.shields.io/github/workflow/status/adamchainz/mariadb-dyncol/CI/main?style=for-the-badge
   :target: https://github.com/adamchainz/mariadb-dyncol/actions?workflow=CI

.. image:: https://img.shields.io/pypi/v/mariadb-dyncol.svg?style=for-the-badge
   :target: https://pypi.org/project/mariadb-dyncol/

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg?style=for-the-badge
   :target: https://github.com/psf/black

.. image:: https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white&style=for-the-badge
   :target: https://github.com/pre-commit/pre-commit
   :alt: pre-commit

Unmaintained (2022-12-07)
-------------------------

I stopped maintaining this package as it has never been popular.
Since MariaDB added JSON support, it’s better to use that for portability, rather than the custom dynamic columns format.
If you’d like to take over maintenance of this package please email me.

----

Pack/unpack Python ``dict``\s into/out of MariaDB's **Dynamic Columns** format.

A quick example:

.. code-block:: pycon

    >>> mariadb_dyncol.pack({"key": "value"})
    b'\x04\x01\x00\x03\x00\x00\x00\x03\x00key!value'
    >>> mariadb_dyncol.unpack(mariadb_dyncol.pack({"key": "value"}))
    {'key': 'value'}

Installation
============

Use **pip**:

.. code-block:: sh

    python -m pip install mariadb-dyncol

Python 3.7 to 3.11 supported.

----

**Working on a Django project?**
Check out my book `Boost Your Django DX <https://adamchainz.gumroad.com/l/byddx>`__ which covers many ways to improve your development experience.

----

Features
========

* Sensible type mapping from Python to SQL
* Tested against examples from MariaDB, including property/fuzz testing with
  `hypothesis <https://hypothesis.readthedocs.io/en/latest/>`_ (which is
  amazing and found many bugs)

Why?
====

The normal way for adding data into dynamic columns fields is with the
``COLUMN_CREATE`` function, and its relatives. This allows you to do things
like:

.. code-block:: sql

    INSERT INTO mytable (attrs) VALUES (COLUMN_CREATE('key', 'value'))

Unfortunately the Django ORM is restricted and cannot use database functions
like this in every instance, at least not until Django 1.9. It was this
limitation I hit whilst implementing a dynamic columns field for my project
`django-mysql <https://github.com/adamchainz/django-mysql>`_ that spurred the
creation of this library.

By pre-packing the dynamic columns, the above query can just insert the blob
of data directly:

.. code-block:: sql

    INSERT INTO mytable (attrs) VALUES (X'0401000300000003006B65792176616C7565')

Asides from being more easily implemented with the Django ORM, this approach
of packing/unpacking dynamic columns in Python also has some advantages:

* All data types are properly preserved in Python. The only way MariaDB
  provides of pulling back all values for a dynamic columns field is to call
  ``COLUMN_JSON``, but JSON only supports strings and integers. Also
  ``COLUMN_JSON`` has a depth limit of 10, but the format has no actual limit.
* The CPU overhead of packing/unpacking the dynamic columns is moved from you
  database server to your (presumably more scalable) clients.

API
===

All functions and names are accessible as attributes of the ``mariadb_dyncol``
module, which you can import with ``import mariadb_dyncol``.

``pack(mapping)``
-----------------

Packs the given mapping (a ``dict``) into the MariaDB Dynamic Columns
format for named columns and returns it as a byte string (Python 3's ``bytes``,
Python 2's ``str``). This is suitable for then inserting into a table as part
of a normal query.

The ``dict``\'s keys must all be unicode strings, and the values must all be
one of the supported data types:

* ``int`` between ``-(2 ** 32) + 1`` and ``(2 ** 64) - 1`` (Python 2: ``long``
  is supported too)
* ``str`` up to 4GB encoded in UTF-8 (Python 2: ``unicode``)
* ``float`` - anything except ``NaN`` or ``+/- inf``
* ``datetime.datetime`` - full range supported
* ``datetime.date`` - full range supported
* ``datetime.time`` - full range supported
* Any ``dict`` that is valid by these rules, allowing nested keys. There is no
  nesting limit except from for MariaDB's ``COLUMN_JSON`` function which
  restricts the depth to 10

Note that this does not support the ``DECIMAL`` type that MariaDB does (and
would naturally map to Python's ``Decimal``) - it is a little more fiddly to
pack/unpack, though certainly possible, and pull requests are welcomed. If you
try and pack a ``Decimal``, a ``DynColNotSupported`` exception will be raised.

There are other restrictions on the UTF-8 encoded column names as documented in
MariaDB:

* The maximum length of a column name is 16383 bytes
* The maximum length of all column names (at one level in nested hierarchies)
  is 65535 bytes

All other unsupported types will raise a ``DynColTypeError``. Out of range
values will raise a ``DynColValueError``.

Examples:

.. code-block:: pycon

    >>> mariadb_dyncol.pack({"a": 1})
    b'\x04\x01\x00\x01\x00\x00\x00\x00\x00a\x02'
    >>> mariadb_dyncol.pack({"a": "💩"})
    b'\x04\x01\x00\x01\x00\x00\x00\x03\x00a!\xf0\x9f\x92\xa9'

``unpack(bytestring)``
----------------------

Unpacks MariaDB dynamic columns data encoded byte string into a dict; the types
you can expect back are those listed above. This is suitable for fetching the
data direct from MariaDB and decoding in Python as opposed to with MariaDB's
``COLUMN_JSON`` function, preserving the types that JSON discards.

As noted above, ``DECIMAL`` values are not supported, and unpacking this
will raise ``DynColNotSupported``. Also strings will only be decoded with the
MySQL charsets ``utf8`` or ``utf8mb4``; strings with other charsets will raise
``DynColNotSupported`` as well.

Unsupported column formats, for example the old MariaDB numbered dynamic
columns format, or corrupt data, will raise ``DynColValueError``.

Examples:

.. code-block:: pycon

    >>> mariadb_dyncol.unpack(b"\x04\x01\x00\x01\x00\x00\x00\x03\x00a!\xf0\x9f\x92\xa9")
    {"a": "💩"}
    >>> mariadb_dyncol.unpack(b"\x04\x01\x00\x01\x00\x00\x00\x00\x00a\x02")
    {"a": 1}


HISTORY.rst
=======
History
=======

3.6.1 (2022-12-08)
------------------

* Mark as unmaintained.

3.6.0 (2022-05-11)
------------------

* Support Python 3.11.

3.5.0 (2022-01-10)
------------------

* Drop Python 3.6 support.

3.4.0 (2021-08-14)
------------------

* Add type hints.

3.3.0 (2021-05-10)
------------------

* Support Python 3.10.

* Stop distributing tests to reduce package size. Tests are not intended to be
  run outside of the tox setup in the repository. Repackagers can use GitHub's
  tarballs per tag.

3.2.0 (2020-12-13)
------------------

* Drop Python 3.5 support.
* Support Python 3.9.
* Move license from BSD to MIT License.

3.1.0 (2019-11-15)
------------------

* Update Python support to 3.5-3.8, as 3.4 has reached its end of life.
* Converted setuptools metadata to configuration file. This meant removing the
  ``__version__`` attribute from the package. If you want to inspect the
  installed version, use
  ``importlib.metadata.version("mariadb-dyncol")``
  (`docs <https://docs.python.org/3.8/library/importlib.metadata.html#distribution-versions>`__ /
  `backport <https://pypi.org/project/importlib-metadata/>`__).

3.0.0 (2019-02-07)
------------------

* Drop Python 2 support, only Python 3.4+ is supported now.

2.0.0 (2018-10-20)
------------------

* Use ``utf8mb4`` character set for encoding strings. This seemed to be broken
  for emoji on older versions of MariaDB (10.1 or 10.2?), so ``utf8`` was
  previously used, however this may have only been a display/``COLUMN_JSON``
  issue on such older versions. MariaDB internally now defaults to ``utf8mb44``
  for dynamic column strings. Since this changes the output of serialization
  slightly, please test before upgrading. Also you probably want to use
  ``utf8mb4`` for everything else MariaDB in your application if you aren't
  already - it is the default on MySQL 8+.

1.2.1 (2017-12-05)
------------------

* Fix a packaging error which caused the tests to be installed alongside the
  package.
* Don't pin version of ``six`` to 1.9.0

1.2.0 (2016-05-24)
------------------

* Disallowed ``str`` values on Python 2 - always use ``unicode``
* Added a benchmark script and made some optimizations that add up to a speed
  boost of about 10%.

1.1.0 (2015-10-13)
------------------

* Tests now verify every operation against MariaDB's ``COLUMN_CHECK`` and
  ``COLUMN_CREATE`` functions
* Fixed column order when >1 UTF8 byte characters are involved
* Fix encoding ``int``\s around size boundaries
* Fix encoding ``time``\s and ``datetime``\s with microseconds=0
* Fix encoding float ``-0.0``
* Fix a data size boundaries off-by-one error
* Fix decoding ``utf8mb4`` strings

1.0.0 (2015-10-09)
------------------

* Support to pack and unpack the named dynamic columns format. No support for
  DECIMAL values or strings with a non utf8mb4 charset.


tests/base.py
from __future__ import annotations

import binascii
import os
from datetime import date
from datetime import datetime
from datetime import time
from typing import Any

import pymysql

from mariadb_dyncol import pack
from mariadb_dyncol import unpack

hexs = binascii.hexlify
unhexs = binascii.unhexlify


def check(
    input: dict[str, Any],
    expected: bytes | None = None,
    expected_prefix: bytes | None = None,
) -> None:
    if expected is not None:
        assert expected_prefix is None

    packed = pack(input)
    assert isinstance(packed, bytes)

    packed_hex = hexs(packed)
    if expected is not None:
        assert packed_hex == expected
    elif expected_prefix is not None:
        packed_hex_prefix = packed_hex[: len(expected_prefix)]
        assert packed_hex_prefix == expected_prefix

    check_against_db(input, packed)

    unpacked = unpack(packed)

    # Nones are not stored and thus we shouldn't compare them
    expected_unpacked = {k: v for k, v in input.items() if v is not None}
    assert unpacked == expected_unpacked


connection = None


def get_connection() -> Any:
    global connection
    if connection is None:
        connection = pymysql.connect(
            host=os.environ.get("MYSQL_HOST", "localhost"),
            password=os.environ["MYSQL_PASSWORD"],
            charset="utf8mb4",
        )
        cursor = connection.cursor()
        try:
            cursor.execute("SELECT VERSION()")
            row = cursor.fetchone()
            assert row is not None
            version = row[0]
        finally:
            cursor.close()
        assert "MariaDB" in version
    return connection


def check_against_db(dicty: dict[str, Any], byte_string: bytes) -> None:
    connection = get_connection()
    cursor = connection.cursor()
    try:
        # Basic validity check
        cursor.execute("SELECT COLUMN_CHECK(%s) AS r", (byte_string,))
        result = cursor.fetchone()[0]
        assert result == 1, f"MariaDB did not validate {hexs(byte_string)!r}"
        # In depth check of re-creating with COLUMN_CREATE
        sql, params = column_create(dicty)
        sql = "SELECT " + sql + " AS v"
        cursor.execute(sql, params)
        result = cursor.fetchone()[0]
        assert hexs(byte_string) == hexs(result)
    finally:
        cursor.close()


def column_create(dicty: dict[str, Any]) -> tuple[str, list[Any]]:
    # COLUMN_CREATE() with no args is invalid so hardcode empty dict
    if not dicty:
        return "COLUMN_DELETE(COLUMN_CREATE('a', 1), 'a')", []

    sql = []
    params: list[Any] = []
    for key, value in dicty.items():
        sql.append("%s")
        params.append(key)
        if isinstance(value, dict):
            subsql, subparams = column_create(value)
            sql.append(subsql)
            params.extend(subparams)
        elif value is None:
            sql.append("NULL")
        elif isinstance(value, (int, str)):
            sql.append("%s")
            params.append(value)
        elif isinstance(value, float):
            # str(float) breaks the query, instead use repr direct as SQL
            sql.append(repr(value) + " AS DOUBLE")
        else:
            sql.append("%s AS " + type_map[type(value)])
            params.append(value)

    return "COLUMN_CREATE(" + ", ".join(sql) + ")", params


type_map = {date: "DATE", datetime: "DATETIME", time: "TIME"}


tests/conftest.py
from __future__ import annotations

import os

from hypothesis import settings

local_settings = settings(max_examples=int(os.getenv("HYPOTHESIS_MAX_EXAMPLES", 100)))
settings.register_profile("fuzztests", local_settings)
settings.load_profile("fuzztests")


tests/fuzztests.py
from math import isinf, isnan

from hypothesis import assume, given
from hypothesis.strategies import dates, datetimes, dictionaries, floats, integers, recursive, text, times

from mariadb_dyncol import DynColValueError, pack, unpack
from mariadb_dyncol.base import MAX_NAME_LENGTH, MAX_TOTAL_NAME_LENGTH  # priv.

from .base import check_against_db

valid_keys = text(
    min_size=1,
    average_size=10.0,
    max_size=MAX_NAME_LENGTH
).filter(
    lambda key: len(key.encode('utf-8')) <= MAX_NAME_LENGTH
)
valid_ints = integers(
    min_value=-(2 ** 31 - 1),
    max_value=(2 ** 64 - 1)
).filter(
    lambda i: abs(i) <= (2 ** 31 - 1) or 0 <= i <= 2 ** 64 - 1
)
valid_floats = floats().filter(
    lambda f: not isnan(f) and not isinf(f)
)
valid_datetimes = datetimes()
valid_dates = dates()
valid_times = times()


def valid_dictionaries(keys, values):
    return dictionaries(keys, values, average_size=5).filter(
        lambda data: (
            sum(len(key.encode('utf-8')) for key in data) <=
            MAX_TOTAL_NAME_LENGTH
        ),
    )


def check_data(data):
    packed = pack(data)
    check_against_db(data, packed)
    assert unpack(packed) == data


@given(valid_dictionaries(valid_keys, valid_ints))
def test_ints(data):
    check_data(data)


@given(valid_dictionaries(valid_keys, valid_floats))
def test_floats(data):
    try:
        check_data(data)
    except DynColValueError:
        assume(False)


@given(valid_dictionaries(valid_keys, text()))
def test_strings(data):
    check_data(data)


@given(valid_dictionaries(valid_keys, valid_datetimes))
def test_datetimes(data):
    check_data(data)


@given(valid_dictionaries(valid_keys, valid_dates))
def test_dates(data):
    check_data(data)


@given(valid_dictionaries(valid_keys, valid_times))
def test_times(data):
    check_data(data)


recursive_values = recursive(
    (valid_ints | valid_floats | text() | valid_datetimes | valid_dates |
     valid_times),
    lambda children: valid_dictionaries(valid_keys, children)
)


@given(valid_dictionaries(valid_keys, recursive_values))
def test_recursively_defined(data):
    try:
        check_data(data)
    except DynColValueError:
        assume(False)


start.py
# -*- coding:utf-8 -*-
from __future__ import unicode_literals

import struct
import unittest

DYN_COL_INT = 0
DYN_COL_DOUBLE = 2
DYN_COL_STRING = 3
DYN_COL_DYNCOL = 8


def column_create(dicty):
    buf = []

    flags = struct.pack('B', 4)
    buf.append(flags)

    column_count = 0
    column_directory = []
    directory_offset = 0
    name_offset = 0
    names = []
    data_offset = 0
    data = []
    for name, value in sorted(dicty.iteritems()):
        if value is None:
            continue

        encname = name.encode('utf-8')
        if isinstance(value, int):
            dtype, encvalue = encode_int(value)
        elif isinstance(value, float):
            dtype, encvalue = encode_float(value)
        elif isinstance(value, basestring):
            dtype, encvalue = encode_string(value)
        elif isinstance(value, dict):
            dtype = DYN_COL_DYNCOL
            encvalue = column_create(value)
        else:
            raise TypeError("Unencodable type {}".format(type(value)))

        column_count += 1
        column_directory.append(struct.pack('H', name_offset))
        column_directory.append(struct.pack('H', (data_offset << 4) + dtype))
        names.append(encname)
        name_offset += len(encname)
        data.append(encvalue)
        data_offset += len(encvalue)

        directory_offset += 2

    buf.append(struct.pack('H', column_count))
    enc_names = b''.join(names)
    buf.append(struct.pack('H', len(enc_names)))
    buf.append(b''.join(column_directory))
    buf.append(enc_names)
    buf.append(b''.join(data))

    return b''.join(buf)


def encode_int(value):
    """
    Stored in the schema:
    0: no data
    -1: 1
     1: 2
    -2: 3
     2: 4
    ...
    """
    if value == 0:
        encoded = b''
    else:
        encoded = abs(2 * value)
        if value < 0:
            encoded -= 1

        cut_last_byte = False
        if encoded <= (2 ** 8 - 1):
            code = 'B'
        elif encoded <= (2 ** 16 - 1):
            code = 'H'
        elif encoded <= (2 ** 24 - 1):
            # Want 3 bytes but only 4 bytes possible with struct
            code = 'I'
            cut_last_byte = True
        elif encoded <= (2 ** 32 - 1):
            code = 'I'
        else:
            raise ValueError("int {} too large".format(value))

        encoded = struct.pack(code, encoded)
        if cut_last_byte:
            encoded = encoded[:-1]
    return DYN_COL_INT, encoded


def encode_float(value):
    return DYN_COL_DOUBLE, struct.pack('d', value)


def encode_string(value):
    encoded = value.encode('utf-8')
    return DYN_COL_STRING, b'\x21' + encoded  # 0x21 = utf8mb4 charset number


def hexs(byte_string):
    return ''.join(("%02X" % ord(x) for x in byte_string))


class ColumnCreateTests(unittest.TestCase):
    def assert_hex(self, dicty, hexstring):
        assert hexs(column_create(dicty)) == hexstring

    def test_a_1(self):
        self.assert_hex({"a": 1}, b"0401000100000000006102")

    def test_a_minus1(self):
        self.assert_hex({"a": -1}, b"0401000100000000006101")

    def test_a_minus2(self):
        self.assert_hex({"a": -2}, b"0401000100000000006103")

    def test_a_0(self):
        self.assert_hex({"a": 0}, b"04010001000000000061")

    def test_a_128(self):
        self.assert_hex({"a": 128}, b"040100010000000000610001")

    def test_a_65535(self):
        self.assert_hex({"a": 65535}, b"04010001000000000061FEFF01")

    def test_a_1048576(self):
        self.assert_hex({"a": 1048576}, b"04010001000000000061000020")

    def test_c_1(self):
        self.assert_hex({"c": 1}, b"0401000100000000006302")

    def test_a_1_b_2(self):
        self.assert_hex(
            {"a": 1, "b": 2},
            b"0402000200000000000100100061620204"
        )

    def test_a_1_b_2_c_3(self):
        self.assert_hex(
            {"a": 1, "b": 2, "c": 3},
            b"0403000300000000000100100002002000616263020406"
        )

    def test_abc_123(self):
        self.assert_hex(
            {"abc": 123},
            b"040100030000000000616263F6"
        )

    def test_string_empty(self):
        self.assert_hex({"a": ""}, b"0401000100000003006121")

    def test_string_values(self):
        self.assert_hex({"a": "string"}, b"0401000100000003006121737472696E67")

    def test_a_unicode_poo(self):
        self.assert_hex({"a": "💩"}, b"0401000100000003006121F09F92A9")

    def test_None(self):
        self.assert_hex({"a": None}, b"0400000000")

    def test_dict(self):
        self.assert_hex(
            {"a": {"b": "c"}},
            b"04010001000000080061040100010000000300622163"
        )

    def test_float_1_0(self):
        self.assert_hex({"a": 1.0}, b"04010001000000020061000000000000F03F")

    def test_float_minus_3_415(self):
        self.assert_hex({"a": -3.415}, b"0401000100000002006152B81E85EB510BC0")

    def test_float_192873409809(self):
        self.assert_hex(
            {"a": 192873409809.0},
            b"040100010000000200610080885613744642"
        )

if __name__ == '__main__':
    unittest.main()


benchmark.py
#!/usr/bin/env python
from __future__ import annotations

import io
import sys
import time
from contextlib import contextmanager
from typing import Any
from typing import Callable
from typing import ContextManager
from typing import Generator
from typing import IO

from tests import base
from tests import test_mariadb_dyncol


def main() -> None:
    nruns = 1000

    def check_against_db(*args: Any, **kwargs: Any) -> None:
        ...

    base.check_against_db = check_against_db

    test_funcs = get_test_funcs()
    print("Running benchmark", file=sys.stderr)
    start = time.time()
    for _ in range(nruns):
        print(".", file=sys.stderr, end="")
        sys.stderr.flush()
        [test_func() for test_func in test_funcs]
    total = time.time() - start
    print("\n", file=sys.stderr)
    print(
        "Ran all tests {} times in {} seconds = {} seconds per run".format(
            nruns, total, total / nruns
        ),
        file=sys.stderr,
    )


def get_test_funcs() -> list[Callable[[], None]]:
    funcs = []
    for name in dir(test_mariadb_dyncol):
        if not name.startswith("test_"):
            continue

        testfunc = getattr(test_mariadb_dyncol, name)
        if hasattr(testfunc, "slow"):
            continue

        if hasattr(testfunc, "skipif") and testfunc.skipif.args[0]:
            continue

        funcs.append(testfunc)
    return funcs


@contextmanager
def captured_output(stream_name: str) -> Generator[IO[Any], None, None]:
    """Return a context manager used by captured_stdout/stdin/stderr
    that temporarily replaces the sys stream *stream_name* with a StringIO.

    Note: This function and the following ``captured_std*`` are copied
          from CPython's ``test.support`` module."""
    orig_stdout = getattr(sys, stream_name)
    setattr(sys, stream_name, io.StringIO())
    try:
        yield getattr(sys, stream_name)
    finally:
        setattr(sys, stream_name, orig_stdout)


def captured_stdout() -> ContextManager[IO[Any]]:
    """Capture the output of sys.stdout:

    with captured_stdout() as stdout:
        print("hello")
    self.assertEqual(stdout.getvalue(), "hello\n")
    """
    return captured_output("stdout")


if __name__ == "__main__":
    main()


src/mariadb_dyncol/base.py
from __future__ import annotations

from datetime import date
from datetime import datetime
from datetime import time
from decimal import Decimal
from math import isinf
from math import isnan
from struct import pack as struct_pack
from struct import unpack as struct_unpack
from struct import unpack_from as struct_unpack_from
from typing import Any
from typing import Callable
from typing import TYPE_CHECKING

if TYPE_CHECKING:  # pragma: no cover
    from typing import NoReturn
else:
    NoReturn = None

DYN_COL_INT = 0
DYN_COL_UINT = 1
DYN_COL_DOUBLE = 2
DYN_COL_STRING = 3
DYN_COL_DECIMAL = 4
DYN_COL_DATETIME = 5
DYN_COL_DATE = 6
DYN_COL_TIME = 7
DYN_COL_DYNCOL = 8

MAX_TOTAL_NAME_LENGTH = 65535
MAX_NAME_LENGTH = MAX_TOTAL_NAME_LENGTH // 4


class DynColLimitError(Exception):
    """
    Indicates that some limit has been reached
    """


class DynColTypeError(TypeError):
    """
    Indicates that a type is wrong
    """


class DynColValueError(ValueError):
    """
    Indicates that a value is wrong
    """


class DynColNotSupported(Exception):
    """
    Indicates a limitation in this implementation
    """


def pack(dicty: dict[str, Any]) -> bytes:
    """
    Convert a mapping into the MariaDB dynamic columns format
    """
    column_count = 0
    column_directory = []
    directory_offset = 0
    name_offset = 0
    names = []
    data_offset = 0
    data = []
    total_encname_length = 0

    dicty_names_encoded = {key.encode("utf-8"): value for key, value in dicty.items()}

    for encname in sorted(dicty_names_encoded.keys(), key=name_order):
        value = dicty_names_encoded[encname]
        if value is None:
            continue

        if len(encname) > MAX_NAME_LENGTH:
            raise DynColLimitError("Key too long: " + encname.decode("utf-8"))
        total_encname_length += len(encname)
        if total_encname_length > MAX_TOTAL_NAME_LENGTH:
            raise DynColLimitError("Total length of keys too long")

        try:
            encode_func = ENCODE_FUNCS[type(value)]
        except KeyError:
            raise DynColTypeError(f"Unencodable type {type(value)}")
        dtype, encvalue = encode_func(value)

        column_count += 1
        column_directory.append(name_offset)
        column_directory.append((data_offset << 4) + dtype)
        names.append(encname)
        name_offset += len(encname)
        data.append(encvalue)
        data_offset += len(encvalue)

        directory_offset += 2

    data_size_flag, coldir_size_code, odd_sized_datacode = data_size(data)

    flags = 4 | data_size_flag  # means this contains named dynamic columns
    enc_names = b"".join(names)

    buf = [struct_pack("<BHH", flags, column_count, len(enc_names))]
    if not odd_sized_datacode:
        buf.append(
            struct_pack(
                "<" + ("H" + coldir_size_code) * (len(column_directory) // 2),
                *column_directory,
            )
        )
    else:
        for i, val in enumerate(column_directory):
            if i % 2 == 0:
                # name_offset
                buf.append(struct_pack("<H", val))
            else:
                # data_offset + dtype, have to cut last byte
                value = struct_pack("<" + coldir_size_code, val)
                buf.append(value[:-1])
    buf.append(enc_names)
    buf.extend(data)
    return b"".join(buf)


def name_order(name: bytes) -> tuple[int, bytes]:
    # Keys are ordered by name length then name
    return len(name), name


def data_size(data: list[bytes]) -> tuple[int, str, bool]:
    data_len = sum(len(d) for d in data)
    if data_len < 0xFFF:
        return 0, "H", False
    elif data_len < 0xFFFFF:
        return 1, "L", True
    elif data_len < 0xFFFFFFF:
        return 2, "L", False
    else:
        raise ValueError("Too much data")


def encode_int(value: int) -> tuple[int, bytes]:
    if value < 0:
        dtype = DYN_COL_INT
        encvalue = -(value << 1) - 1
        if value < -(2**32 - 1):
            raise DynColValueError(f"int {value} out of range")
    else:
        if value <= (2**63 - 1):
            dtype = DYN_COL_INT
            encvalue = value << 1
        elif value <= (2**64 - 1):
            dtype = DYN_COL_UINT
            encvalue = value
        else:
            raise DynColValueError(f"int {value} out of range")

    to_enc = []
    while encvalue:
        to_enc.append(encvalue & 0xFF)
        encvalue = encvalue >> 8
    return dtype, struct_pack("B" * len(to_enc), *to_enc)


def encode_float(value: float) -> tuple[int, bytes]:
    if isnan(value) or isinf(value):
        raise DynColValueError(f"Float value not encodeable: {value}")
    encvalue = struct_pack("d", value)

    # -0.0 is not supported in SQL, change to 0.0
    if encvalue == b"\x00\x00\x00\x00\x00\x00\x00\x80":
        encvalue = b"\x00\x00\x00\x00\x00\x00\x00\x00"

    return DYN_COL_DOUBLE, encvalue


def encode_string(value: str) -> tuple[int, bytes]:
    return DYN_COL_STRING, b"\x2D" + value.encode("utf-8")


def encode_decimal(value: Decimal) -> NoReturn:
    raise DynColNotSupported("Can't encode Decimal values currently")


# def encode_decimal(value):
#     buf = bytearray()
#     intg = int(value)
#     intg_digits = 9
#     buf.extend(struct_pack('>I', intg))

#     frac = value - intg
#     if frac:
#         frac_digits = 1
#         frac_piece = int(str(frac)[2:])  # ugh
#         buf.extend(struct_pack('B', frac_piece))
#     else:
#         frac_digits = 0

#     header = struct_pack('>BB', intg_digits, frac_digits)
#     buf[0] |= 0x80  # Flip the top bit
#     return DYN_COL_DECIMAL, header + bytes(buf)


def encode_datetime(value: datetime) -> tuple[int, bytes]:
    _, enc_date = encode_date(value)
    _, enc_time = encode_time(value)
    return DYN_COL_DATETIME, enc_date + enc_time


def encode_date(value: date) -> tuple[int, bytes]:
    # We don't need any validation since datetime.date is more limited than the
    # MySQL format
    val = value.day | value.month << 5 | value.year << 9
    return DYN_COL_DATE, struct_pack("I", val)[:-1]


def encode_time(value: datetime | time) -> tuple[int, bytes]:
    if value.microsecond > 0:
        val = (
            value.microsecond
            | value.second << 20
            | value.minute << 26
            | value.hour << 32
        )
        return DYN_COL_TIME, struct_pack("Q", val)[:6]
    else:
        val = value.second | value.minute << 6 | value.hour << 12
        return DYN_COL_TIME, struct_pack("I", val)[:3]


def encode_dict(value: dict[str, Any]) -> tuple[int, bytes]:
    return DYN_COL_DYNCOL, pack(value)


ENCODE_FUNCS: dict[type[Any], Callable[[Any], tuple[int, bytes]]] = {
    int: encode_int,
    date: encode_date,
    datetime: encode_datetime,
    time: encode_time,
    float: encode_float,
    str: encode_string,
    Decimal: encode_decimal,
    dict: encode_dict,
}


def unpack(buf: bytes) -> dict[str, Any]:
    """
    Convert MariaDB dynamic columns data in a byte string into a dict
    """
    flags: int
    column_count: int
    len_names: int
    flags, column_count, len_names = struct_unpack_from("<BHH", buf)
    data_offset_code, coldata_size, data_offset_mask = decode_data_size(flags)
    if (flags & 0xFC) != 4:
        raise DynColValueError("Unknown dynamic columns format")

    if column_count == 0:
        return {}

    column_directory_end = (1 + 2 + 2) + coldata_size * column_count
    names_end = column_directory_end + len_names

    column_directory = buf[1 + 2 + 2 : column_directory_end]
    enc_names = buf[column_directory_end:names_end]
    data = buf[names_end:]

    names = {}
    values = {}

    last_name_offset: int | None = None
    last_data_offset: int | None = None
    last_dtype: int | None = None
    name_offset: int
    data_offset_dtype: int

    for i in range(column_count):
        if coldata_size % 2 == 0:
            name_offset, data_offset_dtype = struct_unpack_from(
                "<H" + data_offset_code, column_directory, offset=i * coldata_size
            )
        else:
            (name_offset,) = struct_unpack_from(
                "<H", column_directory, offset=i * coldata_size
            )
            # can't struct_unpack the 3 bytes so hack around
            dodt_bytes = (
                column_directory[i * coldata_size + 2 : (i * coldata_size + 5)]
                + b"\x00"
            )
            (data_offset_dtype,) = struct_unpack("<" + data_offset_code, dodt_bytes)

        data_offset_dtype &= data_offset_mask
        data_offset = data_offset_dtype >> 4
        dtype = data_offset_dtype & 0xF

        # Store *last* column's name
        if last_name_offset is not None:
            names[i - 1] = enc_names[last_name_offset:name_offset].decode("utf-8")
        last_name_offset = name_offset

        #
        if last_data_offset is not None:
            assert last_dtype is not None
            values[i - 1] = decode(last_dtype, data[last_data_offset:data_offset])
        last_data_offset = data_offset
        last_dtype = dtype

    names[column_count - 1] = enc_names[last_name_offset:].decode("utf-8")
    assert last_dtype is not None
    values[column_count - 1] = decode(last_dtype, data[last_data_offset:])

    # join data and names
    return {names[i]: values[i] for i in range(column_count)}


def decode_data_size(flags: int) -> tuple[str, int, int]:
    t = flags & 0x03
    if t == 0:
        return "H", 4, 0xFFFF
    elif t == 1:
        return "L", 5, 0xFFFFFF
    elif t == 2:
        return "L", 6, 0xFFFFFFFF
    else:
        raise ValueError("Unknown dynamic columns format")


def decode(dtype: int, encvalue: bytes) -> Any:
    try:
        decode_func = DECODE_FUNCS[dtype]
    except KeyError:
        raise ValueError()
    return decode_func(encvalue)


def decode_int(encvalue: bytes) -> int:
    value = 0
    for i, b in enumerate(bytearray(encvalue)):
        value += b << (8 * i)
    if value & 1:
        return -(value >> 1) - 1
    else:
        return value >> 1


def decode_uint(encvalue: bytes) -> int:
    value: int
    (value,) = struct_unpack("Q", encvalue)
    return value


def decode_double(encvalue: bytes) -> float:
    value: float
    (value,) = struct_unpack("d", encvalue)
    return value


def decode_string(encvalue: bytes) -> str:
    if not encvalue.startswith((b"\x21", b"\x2D")):
        raise DynColNotSupported(
            "Can only decode strings with MySQL charsets utf8 or utf8mb4"
        )
    return encvalue[1:].decode("utf-8")


def decode_decimal(encvalue: bytes) -> NoReturn:
    raise DynColNotSupported("Can't decode Decimal values currently")


# def decode_decimal(encvalue):
#     num_intg, num_frac = struct_unpack('>BB', encvalue[:2])
#     intg, = struct_unpack('>I', encvalue[2:6])
#     intg ^= 0x80000000
#     if num_frac == 0:
#         frac = 0
#     else:
#         frac, = struct_unpack('>B', encvalue[6:])
#     return Decimal(str(intg) + '.' + str(frac))


def decode_datetime(encvalue: bytes) -> datetime:
    d = decode_date(encvalue[:3])
    t = decode_time(encvalue[3:])
    return datetime.combine(d, t)


def decode_date(encvalue: bytes) -> date:
    (val,) = struct_unpack("I", encvalue + b"\x00")
    return date(day=val & 0x1F, month=(val >> 5) & 0xF, year=(val >> 9))


def decode_time(encvalue: bytes) -> time:
    if len(encvalue) == 6:
        (val,) = struct_unpack("Q", encvalue + b"\x00\x00")
        return time(
            microsecond=val & 0xFFFFF,
            second=(val >> 20) & 0x3F,
            minute=(val >> 26) & 0x3F,
            hour=(val >> 32),
        )
    else:  # must be 3
        (val,) = struct_unpack("I", encvalue + b"\x00")
        return time(
            microsecond=0,
            second=(val) & 0x3F,
            minute=(val >> 6) & 0x3F,
            hour=(val >> 12),
        )


DECODE_FUNCS: dict[int, Callable[[bytes], Any]] = {
    DYN_COL_INT: decode_int,
    DYN_COL_UINT: decode_uint,
    DYN_COL_DOUBLE: decode_double,
    DYN_COL_STRING: decode_string,
    DYN_COL_DECIMAL: decode_decimal,
    DYN_COL_DATETIME: decode_datetime,
    DYN_COL_DATE: decode_date,
    DYN_COL_TIME: decode_time,
    DYN_COL_DYNCOL: unpack,
}
